---
title: "Integration with other panels"
author: 
  - name: Kevin Rue-Albrecht
    affiliation:
    - University of Oxford
    email: kevin.rue-albrecht@imm.ox.ac.uk
output: 
  BiocStyle::html_document:
    self_contained: yes
    toc: true
    toc_float: true
    toc_depth: 2
    code_folding: show
date: "`r doc_date()`"
package: "`r pkg_ver('iSEEpathways')`"
vignette: >
  %\VignetteIndexEntry{Integration with other panels}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>",
    crop = NULL ## Related to https://stat.ethz.ch/pipermail/bioc-devel/2020-April/016656.html
)
```

```{r, eval=!exists("SCREENSHOT"), include=FALSE}
SCREENSHOT <- function(x, ...) knitr::include_graphics(x)
```

```{r vignetteSetup, echo=FALSE, message=FALSE, warning = FALSE}
## Track time spent on making the vignette
startTime <- Sys.time()

## Bib setup
library("RefManageR")

## Write bibliography information
bib <- c(
    R = citation(),
    BiocStyle = citation("BiocStyle")[1],
    knitr = citation("knitr")[1],
    RefManageR = citation("RefManageR")[1],
    rmarkdown = citation("rmarkdown")[1],
    sessioninfo = citation("sessioninfo")[1],
    testthat = citation("testthat")[1],
    iSEEpathways = citation("iSEEpathways")[1]
)
```

# Scenario

In this vignette, we use the example of a differential expression and pathway analysis workflow to demonstrate how a selection made in a panel of pathway analysis results may be transmitted to other row-oriented panels in the `r BiocStyle::Biocpkg("iSEE")` application.

# Example data

We use the `?airway` data set.

We briefly adjust the reference level of the treatment factor to the untreated condition.

```{r, message=FALSE, warning=FALSE}
library("airway")
data("airway")
airway$dex <- relevel(airway$dex, "untrt")
```

We also map the Ensembl gene identifiers to more recognisable gene symbols, setting `?rownames` to a unique identifier composed of either gene symbol, gene identifier, of a concatenate of both.

```{r, message=FALSE, warning=FALSE}
library("org.Hs.eg.db")
library("scater")
rowData(airway)[["ENSEMBL"]] <- rownames(airway)
rowData(airway)[["SYMBOL"]] <- mapIds(org.Hs.eg.db, rownames(airway), "SYMBOL", "ENSEMBL")
rownames(airway) <- uniquifyFeatureNames(
  ID = rowData(airway)[["ENSEMBL"]],
  names = rowData(airway)[["SYMBOL"]]
)
```

We also compute log-counts for better visualisation of differential expression in the `r BiocStyle::Biocpkg("iSEE")` app.

```{r}
library("scuttle")
airway <- logNormCounts(airway)
```

We run a standard Limma-Voom analysis using `limma::voom()`, `limma::lmFit()`, `limma::makeContrasts()`, and `limma::eBayes()`.

```{r}
library("edgeR")

counts <- assay(airway, "counts")
design <- model.matrix(~ 0 + dex + cell, data = colData(airway))

keep <- filterByExpr(counts, design)
v <- voom(counts[keep,], design, plot=FALSE)
fit <- lmFit(v, design)
contr <- makeContrasts("dextrt - dexuntrt", levels = colnames(coef(fit)))
tmp <- contrasts.fit(fit, contr)
tmp <- eBayes(tmp)
res_limma <- topTable(tmp, sort.by = "P", n = Inf)
head(res_limma)
```

Then, we embed this set of differential expression results in the `?airway` object using the `embedResults()` method.

```{r}
library("iSEEde")
airway <- embedResults(res_limma, airway, name = "Limma-Voom", class = "limma")
rowData(airway)
```

We prepare Gene Ontology gene sets using `r BiocStyle::Biocpkg("org.Hs.eg.db")`.

```{r}
library("org.Hs.eg.db")
pathways <- select(org.Hs.eg.db, keys(org.Hs.eg.db, "SYMBOL"), c("GOALL"), keytype = "SYMBOL")
pathways <- subset(pathways, ONTOLOGYALL == "BP")
pathways <- unique(pathways[, c("SYMBOL", "GOALL")])
pathways <- split(pathways$SYMBOL, pathways$GOALL)
```

Separately, we define and register a function that fetches the gene identifiers associated with a given pathway identifier.
The function definition must include an ellipsis to accommodate optional arguments passed to other variants of the function.
For instance, see section [Trading off memory usage for speed](#speed-mapping).

```{r}
map_GO <- function(pathway_id, ...) {
    select(org.Hs.eg.db::org.Hs.eg.db, pathway_id, "SYMBOL", keytype = "GOALL")[["SYMBOL"]]
}
airway <- registerAppOptions(airway, Pathways.map.functions = list(GO = map_GO))
```

We run a standard GSEA analysis using `r BiocStyle::Biocpkg("fgsea")`.

```{r}
library("fgsea")
set.seed(42)
fgseaRes <- fgsea(pathways = pathways, 
                  stats    = na.omit(log2FoldChange(rowData(airway)[["iSEEde"]][["Limma-Voom"]])),
                  minSize  = 15,
                  maxSize  = 500)
head(fgseaRes[order(pval), ])
```

We embed this set of pathway analysis results in the airway object (manually for now).

But first, we reorder the results by increasing p-value, convert the `?data.table` to a `?data.frame`, and set the `rownames` to the pathway identifier.

```{r}
library("iSEEpathways")
fgseaRes <- fgseaRes[order(pval), ]
fgseaRes <- as.data.frame(fgseaRes)
rownames(fgseaRes) <- fgseaRes$pathway
airway <- embedPathwaysResults(fgseaRes, airway, name = "fgsea", class = "fgsea")
airway
```

We define and register a function that displays details for the selected Gene Ontology gene set.

```{r}
library("GO.db")
library("shiny")
library("iSEE")
go_details <- function(x) {
    info <- select(GO.db, x, c("TERM", "ONTOLOGY", "DEFINITION"), "GOID")
    html <- list(p(strong(info$GOID), ":", info$TERM, paste0("(", info$ONTOLOGY, ")")))
    if (!is.na(info$DEFINITION)) {
        html <- append(html, list(p(info$DEFINITION)))
    }
    tagList(html)
}
airway <- registerAppOptions(airway, PathwaysTable.select.details = go_details)
```


```{r "start", message=FALSE, warning=FALSE}
app <- iSEE(airway, initial = list(
  PathwaysTable(ResultName="fgsea", PanelWidth = 4L),
  VolcanoPlot(RowSelectionSource = "PathwaysTable1", PanelWidth = 4L),
  ComplexHeatmapPlot(RowSelectionSource = "PathwaysTable1",
      PanelWidth = 4L, PanelHeight = 700L,
      CustomRows = FALSE, ColumnData = "dex",
      ClusterRows = TRUE, ClusterRowsDistance = "euclidean", AssayCenterRows = TRUE)
))

if (interactive()) {
  shiny::runApp(app)
}
```

```{r, echo=FALSE, out.width="100%"}
SCREENSHOT("screenshots/integration.png", delay=30)
```


# Trading off memory usage for speed {#speed-mapping}

The function `map_GO()` that we defined earlier to map a pathway identifier to a set of gene identifiers uses the `r BiocStyle::Biocpkg("org.Hs.eg.db")` package and the `?AnnotationDbi::select()` function.

While memory-efficient, the repeated calls to the database of gene annotations introduce a bottleneck that limits the reactivity of the app.

We can restore the speed of the app rendering by trading off an increased memory usage.
Specifically, we can use the object `pathways` that we also created earlier as a named list of pathway identifiers and character vectors of gene identifiers associated with each pathway.

First, for the app to be self-contained, the list of pathways should be stored within the `airway` object itself.

```{r}
metadata(airway)[["pathways"]] <- list(GO = pathways)
```

Then, we can write a new faster function that fetches gene identifiers directly from that list rather than the database.
The function should take as first argument a single pathway identifier, and the second argument must be called `se` to match the name of the `SummarizedExperiment` object used within the app.
As a a trade off, the app now relies on the list of pathway annotations being available as an object in the <i class="fab fa-r-project"></i> session, using additional memory for the benefit of dramatically faster access.

```{r}
map_GO_v2 <- function(pathway_id, se) {
    metadata(se)[["pathways"]][["GO"]][[pathway_id]]
}
airway <- registerAppOptions(airway, Pathways.map.functions = list(GO = map_GO_v2))
```

We can then launch a new instance of the app, using the same initial configuration, but the update `airway` object.

```{r "faster", message=FALSE, warning=FALSE}
app <- iSEE(airway, initial = list(
  PathwaysTable(ResultName="fgsea", PanelWidth = 4L),
  VolcanoPlot(RowSelectionSource = "PathwaysTable1", PanelWidth = 4L),
  ComplexHeatmapPlot(RowSelectionSource = "PathwaysTable1",
      PanelWidth = 4L, PanelHeight = 700L,
      CustomRows = FALSE, ColumnData = "dex",
      ClusterRows = TRUE, ClusterRowsDistance = "euclidean", AssayCenterRows = TRUE)
))

if (interactive()) {
  shiny::runApp(app)
}
```

```{r, echo=FALSE, out.width="100%"}
SCREENSHOT("screenshots/integration-faster.png", delay=20)
```


# Reproducibility

The `r Biocpkg("iSEEpathways")` package `r Citep(bib[["iSEEpathways"]])` was made possible thanks to:

* R `r Citep(bib[["R"]])`
* `r Biocpkg("BiocStyle")` `r Citep(bib[["BiocStyle"]])`
* `r CRANpkg("knitr")` `r Citep(bib[["knitr"]])`
* `r CRANpkg("RefManageR")` `r Citep(bib[["RefManageR"]])`
* `r CRANpkg("rmarkdown")` `r Citep(bib[["rmarkdown"]])`
* `r CRANpkg("sessioninfo")` `r Citep(bib[["sessioninfo"]])`
* `r CRANpkg("testthat")` `r Citep(bib[["testthat"]])`

This package was developed using `r BiocStyle::Biocpkg("biocthis")`.


Code for creating the vignette

```{r createVignette, eval=FALSE}
## Create the vignette
library("rmarkdown")
system.time(render("integration.Rmd", "BiocStyle::html_document"))

## Extract the R code
library("knitr")
knit("integration.Rmd", tangle = TRUE)
```

Date the vignette was generated.

```{r reproduce1, echo=FALSE}
## Date the vignette was generated
Sys.time()
```

Wallclock time spent generating the vignette.

```{r reproduce2, echo=FALSE}
## Processing time in seconds
totalTime <- diff(c(startTime, Sys.time()))
round(totalTime, digits = 3)
```

`R` session information.

```{r reproduce3, echo=FALSE}
## Session info
library("sessioninfo")
options(width = 120)
session_info()
```


# Bibliography

This vignette was generated using `r Biocpkg("BiocStyle")` `r Citep(bib[["BiocStyle"]])`
with `r CRANpkg("knitr")` `r Citep(bib[["knitr"]])` and `r CRANpkg("rmarkdown")` `r Citep(bib[["rmarkdown"]])` running behind the scenes.

Citations made with `r CRANpkg("RefManageR")` `r Citep(bib[["RefManageR"]])`.

```{r vignetteBiblio, results = "asis", echo = FALSE, warning = FALSE, message = FALSE}
## Print bibliography
PrintBibliography(bib, .opts = list(hyperlink = "to.doc", style = "html"))
```
